packages/components/src/color-picker/inputs.js

```diff
@@ -0,0 +1,287 @@
+		onChange( { [ propName ]: value } );
+	}
+
+	handleChange( value ) {
+		const { label, onChange } = this.props;
+		// Protect against expanding a value while we're typing.
+		if ( value.length > 4 ) {
+			const propName = String( label ).toLowerCase();
+			onChange( { [ propName ]: value } );
+		}
+		this.setState( { value } );
+	}
+
+	handleKeyDown( { keyCode } ) {
+		if ( keyCode !== ENTER && keyCode !== UP && keyCode !== DOWN ) {
+			return;
+		}
+		const { value } = this.state;
+		const { label, onChange } = this.props;
+		const propName = String( label ).toLowerCase();
+		onChange( { [ propName ]: value } );
+	}
+
+	render() {
+		const { label, ...props } = this.props;
+		const { value } = this.state;
+		return (
+			<TextControl
+				className="components-color-picker__inputs-field"
+				label={ label }
+				value={ value }
+				onChange={ ( newValue ) => this.handleChange( newValue ) }
+				onBlur={ this.handleBlur }
+				onKeyDown={ this.handleKeyDown }
+				{ ...omit( props, [ 'onChange', 'value' ] ) }
+			/>
+		);
+	}
+}
+
+export class Inputs extends Component {
+	constructor( { hsl } ) {
+		super( ...arguments );
+
+		const view = hsl.a === 1 ? 'hex' : 'rgb';
+		this.state = { view };
+
+		this.toggleViews = this.toggleViews.bind( this );
+		this.handleChange = this.handleChange.bind( this );
+	}
+
+	static getDerivedStateFromProps( props, state ) {
+		if ( props.hsl.a !== 1 && state.view === 'hex' ) {
+			return { view: 'rgb' };
+		}
+		return null;
+	}
+
+	toggleViews() {
+		if ( this.state.view === 'hex' ) {
+			this.setState( { view: 'rgb' } );
+		} else if ( this.state.view === 'rgb' ) {
+			this.setState( { view: 'hsl' } );
+		} else if ( this.state.view === 'hsl' ) {
+			if ( this.props.hsl.a === 1 ) {
+				this.setState( { view: 'hex' } );
+			} else {
+				this.setState( { view: 'rgb' } );
+			}
+		}
+	}
+
+	handleChange( data ) {
+		if ( data.hex ) {
+			if ( isValidHex( data.hex ) ) {
+				this.props.onChange( {
+					hex: data.hex,
+					source: 'hex',
+				} );
+			}
+		} else if ( data.r || data.g || data.b ) {
+			this.props.onChange( {
+				r: data.r || this.props.rgb.r,
+				g: data.g || this.props.rgb.g,
+				b: data.b || this.props.rgb.b,
+				source: 'rgb',
+			} );
+		} else if ( data.a ) {
+			if ( data.a < 0 ) {
+				data.a = 0;
+			} else if ( data.a > 1 ) {
+				data.a = 1;
+			}
+
+			this.props.onChange( {
+				h: this.props.hsl.h,
+				s: this.props.hsl.s,
+				l: this.props.hsl.l,
+				a: Math.round( data.a * 100 ) / 100,
+				source: 'rgb',
+			} );
+		} else if ( data.h || data.s || data.l ) {
+			// Remove any occurances of '%'.
+			if ( typeof data.s === 'string' && data.s.includes( '%' ) ) {
+				data.s = data.s.replace( '%', '' );
+			}
+			if ( typeof data.l === 'string' && data.l.includes( '%' ) ) {
+				data.l = data.l.replace( '%', '' );
+			}
+
+			this.props.onChange( {
+				h: data.h || this.props.hsl.h,
+				s: Number( ( data.s && data.s ) || this.props.hsl.s ),
```

Comment:
It can be, so this should account for that.

Reply:
Also to note: these values are taken from the `input type=number` component and can be void. The same for the RGB code above.

Reply:
`data.s` will always be a string. It can be `"0"` which is a truthy value. So, the only case where the `this.props.s` is used is when `data.s` is blank.